How to access HDFS file using "gdal.Open"? The method of "/vsihdfs/hdfs://" has failed.

Hi,
    I've uploaded a ".tif" image to HDFS cluster, and it can be accessed by "hdfs dfs -ls hdfs://rsMaster:9000/NC_H08_20180715_1230_R21_FLDK.06001_06001.tif".
-rw-r--r--   2 yunyao yunyao      78058 2023-04-06 14:42 hdfs://rsMaster:9000/NC_H08_20180715_1230_R21_FLDK.06001_06001.tif
    However, I'm unable to access it using following codes:    

from osgeo import gdal, osr, ogr, gdalconst

inpath = "/vsihdfs/hdfs://rsMaster:9000/NC_H08_20180715_1230_R21_FLDK.06001_06001.tif"
ds = gdal.Open(inpath)

band = ds.RasterCount  # 波段数
geoTransform = ds.GetGeoTransform()  # 坐标系
proj = ds.GetProjection()  # 投影

hdfsListDirectory(hdfs://rsMaster:9000): FileSystem#listStatus error:
IllegalArgumentException: Pathname  from hdfs://rsMaster:9000 is not a valid DFS filename.java.lang.IllegalArgumentException: Pathname  from hdfs://rsMaster:9000 is not a valid DFS filename.
	at org.apache.hadoop.hdfs.DistributedFileSystem.getPathName(DistributedFileSystem.java:243)
	at org.apache.hadoop.hdfs.DistributedFileSystem.listStatusInternal(DistributedFileSystem.java:1052)
	at org.apache.hadoop.hdfs.DistributedFileSystem.access$1000(DistributedFileSystem.java:131)
	at org.apache.hadoop.hdfs.DistributedFileSystem$24.doCall(DistributedFileSystem.java:1119)
	at org.apache.hadoop.hdfs.DistributedFileSystem$24.doCall(DistributedFileSystem.java:1116)
	at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81)
	at org.apache.hadoop.hdfs.DistributedFileSystem.listStatus(DistributedFileSystem.java:1126)

    So, could you pls help to give me some suggestions? Thanks!

Hi, the issue still exist! When I type "ds = gdal.Open("/vsihdfs/hdfs:/NC_H08_20180715_1230_R21_FLDK.06001_06001.tif")", the line can be executed. Also, "ds.RasterCount" and "proj = ds.GetProjection()" can print the result data.
>>> import gdal
>>> from osgeo import gdal_array, osr

>>> ds = gdal.Open("/vsihdfs/hdfs:/NC_H08_20180715_1230_R21_FLDK.06001_06001.tif")

>>> band = ds.RasterCount
>>> band
21

>>> proj = ds.GetProjection() 
>>> proj
'GEOGCS["WGS 84",DATUM["WGS_1984",SPHEROID["WGS 84",6378137,298.257223563,AUTHORITY["EPSG","7030"]],AUTHORITY["EPSG","6326"]],PRIMEM["Greenwich",0],UNIT["degree",0.0174532925199433],AUTHORITY["EPSG","4326"]]'

    However, when I run "proj = ds.GetProjection()", the following errors appear:
>>> proj = ds.GetProjection()
hdfsListDirectory(hdfs:): constructNewObjectOfPath error:
URISyntaxException: Expected scheme-specific part at index 5: hdfs:java.lang.IllegalArgumentException: java.net.URISyntaxException: Expected scheme-specific part at index 5: hdfs:
	at org.apache.hadoop.fs.Path.initialize(Path.java:259)
	at org.apache.hadoop.fs.Path.<init>(Path.java:217)
Caused by: java.net.URISyntaxException: Expected scheme-specific part at index 5: hdfs:
	at java.net.URI$Parser.fail(URI.java:2847)
	at java.net.URI$Parser.failExpecting(URI.java:2853)
	at java.net.URI$Parser.parse(URI.java:3056)
	at java.net.URI.<init>(URI.java:746)
	at org.apache.hadoop.fs.Path.initialize(Path.java:256)
	... 1 more

    So, could you pls help to give me some suggestions? Thanks!

--- a/autotest/gcore/vsihdfs.py
+++ b/autotest/gcore/vsihdfs.py
@@ -37,6 +37,18 @@ import pytest
 
 from osgeo import gdal
 
+"""
+Those tests require $HADOOP_ROOT/etc/hadoop/core-site.xml with a configuration
+allowing file:// access:
+
+<configuration>
+    <property>
+        <name>fs.defaultFS</name>
+        <value>file:///</value>
+        <!-- <value>hdfs://localhost:9000</value> -->
+    </property>
+</configuration>
+"""
 
 # Read test
 def test_vsihdfs_1():
@@ -100,11 +112,6 @@ def test_vsihdfs_3():
     gdal.VSIFCloseL(fp)
 
 
-# Write test
-def test_vsihdfs_4():
-    pytest.skip()
-
-
 # EOF test
 def test_vsihdfs_5():
     if gdaltest.have_vsihdfs == False:
@@ -120,15 +127,7 @@ def test_vsihdfs_5():
 
     gdal.VSIFReadL(1000000, 1, fp)
     eof = gdal.VSIFEofL(fp)
-    assert eof == 0
-
-    gdal.VSIFReadL(1, 1, fp)
-    eof = gdal.VSIFEofL(fp)
     assert eof == 1
-
-    gdal.VSIFSeekL(fp, 0, 0)
-    eof = gdal.VSIFEofL(fp)
-    assert eof == 0
 
     gdal.VSIFCloseL(fp)
 

--- a/doc/source/development/building_from_source.rst
+++ b/doc/source/development/building_from_source.rst
@@ -753,6 +753,8 @@ detect the HDF5 library.
     be manually set when needed.
 
 
+.. _building_from_source_hdfs:
+
 HDFS
 ****
 

--- a/doc/source/user/virtual_file_systems.rst
+++ b/doc/source/user/virtual_file_systems.rst
@@ -656,7 +656,32 @@ Authentication options, and read-only features, are identical to :ref:`/vsiswift
 /vsihdfs/ (Hadoop File System)
 ++++++++++++++++++++++++++++++
 
-/vsihdfs/ is a file system handler that provides read access to HDFS. This handler requires GDAL to have been built with Java support (``--with-java``) and HDFS support (``--with-hdfs``). Support for this handler is currently only available on Unix-like systems. Note: support for the HTTP REST API (webHdfs) is also available with :ref:`vsiwebhdfs`
+/vsihdfs/ is a file system handler that provides read access to HDFS.
+This handler requires GDAL to have been built with Java support
+(CMake `FindJNI <https://cmake.org/cmake/help/latest/module/FindJNI.html>`__)
+and :ref:`HDFS <building_from_source_hdfs>` support.
+Support for this handler is currently only available on Unix-like systems.
+
+Note: support for the HTTP REST API (webHdfs) is also available with :ref:`vsiwebhdfs`
+
+The LD_LIBRARY_PATH and CLASSPATH environment variables must be typically
+set up as following.
+
+::
+
+    HADOOP_HOME=$HOME/hadoop-3.3.5
+    LD_LIBRARY_PATH=$HADOOP_HOME/lib/native:$LD_LIBRARY_PATH
+    CLASSPATH=$HADOOP_HOME/etc/hadoop:$HADOOP_HOME/share/hadoop/common/*:$HADOOP_HOME/share/hadoop/common/lib/*:$HADOOP_HOME/share/hadoop/hdfs/*
+
+
+Failure to properly define the CLASSPATH will result in hard crashes in the
+native libhdfs.
+
+Relevant Hadoop documentation links:
+
+- `C API libhdfs <https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-hdfs/LibHdfs.html>`__
+- `HDFS Users Guide <https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-hdfs/HdfsUserGuide.html>`__
+- `Hadoop: Setting up a Single Node Cluster <https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-common/SingleCluster.html>`__
 
 Recognized filenames are of the form :file:`/vsihdfs/hdfsUri` where ``hdfsUri`` is a valid HDFS URI.
 
@@ -664,8 +689,8 @@ Examples:
 
 ::
 
-    /vsihdfs/file:/tmp/my.tif  (a local file accessed through HDFS)
-    /vsihdfs/hdfs:/hadoop/my.tif  (a file stored in HDFS)
+    /vsihdfs/file:/home/user//my.tif  (a local file accessed through HDFS)
+    /vsihdfs/hdfs://localhost:9000/my.tif  (a file stored in HDFS)
 
 .. versionadded:: 2.4
 

--- a/port/CMakeLists.txt
+++ b/port/CMakeLists.txt
@@ -261,11 +261,6 @@ if (HAVE_USERFAULTFD_H)
   target_compile_definitions(cpl PRIVATE -DENABLE_UFFD)
 endif ()
 
-option(HDFS_ENABLED "Include HDFS support" OFF)
-if (HDFS_ENABLED)
-  target_compile_definitions(cpl PRIVATE -DHDFS_ENABLED)
-endif ()
-
 # for plugin DLFCN: for win32 https://github.com/dlfcn-win32/dlfcn-win32/archive/v1.1.1.tar.gz if(WIN32)
 # find_package(dlfcn- win32 REQUIRED) set(CMAKE_DL_LIBS dlfcn-win32::dl) endif()
 

--- a/port/cpl_vsil_hdfs.cpp
+++ b/port/cpl_vsil_hdfs.cpp
@@ -93,11 +93,8 @@ class VSIHdfsHandle final : public VSIVirtualHandle
     bool bEOF = false;
 
   public:
-#if __cplusplus >= 201103L
     static constexpr const char *VSIHDFS = "/vsihdfs/";
-#else
-    static const char *VSIHDFS = "/vsihdfs/";
-#endif
+
     VSIHdfsHandle(hdfsFile poFile, hdfsFS poFilesystem, const char *pszFilename,
                   bool bReadOnly);
     ~VSIHdfsHandle() override;
@@ -169,6 +166,8 @@ size_t VSIHdfsHandle::Read(void *pBuffer, size_t nSize, size_t nMemb)
 
         if (bytes > 0)
         {
+            if (static_cast<size_t>(bytes) < bytes_to_request)
+                bEOF = true;
             bytes_read += bytes;
         }
         if (bytes == 0)
@@ -311,13 +310,23 @@ VSIHdfsFilesystemHandler::Open(const char *pszFilename, const char *pszAccess,
     return nullptr;
 }
 
-int VSIHdfsFilesystemHandler::Stat(const char *pszeFilename,
+int VSIHdfsFilesystemHandler::Stat(const char *pszFilename,
                                    VSIStatBufL *pStatBuf, int)
 {
+    memset(pStatBuf, 0, sizeof(VSIStatBufL));
+
+    if (strncmp(pszFilename, VSIHdfsHandle::VSIHDFS,
+                strlen(VSIHdfsHandle::VSIHDFS)) != 0)
+    {
+        return -1;
+    }
+
     EnsureFilesystem();
 
+    // CPLDebug("VSIHDFS", "Stat(%s)", pszFilename);
+
     hdfsFileInfo *poInfo = hdfsGetPathInfo(
-        poFilesystem, pszeFilename + strlen(VSIHdfsHandle::VSIHDFS));
+        poFilesystem, pszFilename + strlen(VSIHdfsHandle::VSIHDFS));
 
     if (poInfo != nullptr)
     {
@@ -381,21 +390,58 @@ int VSIHdfsFilesystemHandler::Rmdir(const char *)
 
 char **VSIHdfsFilesystemHandler::ReadDir(const char *pszDirname)
 {
+    if (strncmp(pszDirname, VSIHdfsHandle::VSIHDFS,
+                strlen(VSIHdfsHandle::VSIHDFS)) != 0)
+    {
+        return nullptr;
+    }
+
     EnsureFilesystem();
 
-    int mEntries = 0;
+    std::string osDirName(pszDirname);
+    if (osDirName.back() != '/')
+        osDirName += '/';
+
+    VSIStatBufL sStat;
+    if (Stat(osDirName.c_str(), &sStat, 0) != 0 || sStat.st_mode != S_IFDIR)
+        return nullptr;
+
+    int nEntries = 0;
+    std::string osDirNameWithoutPrefix(
+        osDirName.substr(strlen(VSIHdfsHandle::VSIHDFS)));
+
+    // file:///home/user/... is accepted, but if this is used, files returned
+    // by hdfsListDirectory() use file:/home/user/...
+    if (osDirNameWithoutPrefix.compare(0, strlen("file:///"), "file:///") == 0)
+    {
+        osDirNameWithoutPrefix =
+            "file:/" + osDirNameWithoutPrefix.substr(strlen("file:///"));
+    }
+
     hdfsFileInfo *paoInfo = hdfsListDirectory(
-        poFilesystem, pszDirname + strlen(VSIHdfsHandle::VSIHDFS), &mEntries);
-    char **retval = nullptr;
+        poFilesystem, osDirNameWithoutPrefix.c_str(), &nEntries);
 
     if (paoInfo != nullptr)
     {
         CPLStringList aosNames;
-        for (int i = 0; i < mEntries; ++i)
-            aosNames.AddString(paoInfo[i].mName);
-        retval = aosNames.StealList();
-        hdfsFreeFileInfo(paoInfo, mEntries);
-        return retval;
+        for (int i = 0; i < nEntries; ++i)
+        {
+            // CPLDebug("VSIHDFS", "[%d]: %s", i, paoInfo[i].mName);
+            if (STARTS_WITH(paoInfo[i].mName, osDirNameWithoutPrefix.c_str()))
+            {
+                aosNames.AddString(paoInfo[i].mName +
+                                   osDirNameWithoutPrefix.size());
+            }
+            else
+            {
+                CPLDebug("VSIHDFS",
+                         "hdfsListDirectory() returned %s, but this is not "
+                         "starting with %s",
+                         paoInfo[i].mName, osDirNameWithoutPrefix.c_str());
+            }
+        }
+        hdfsFreeFileInfo(paoInfo, nEntries);
+        return aosNames.StealList();
     }
     return nullptr;
 }

